### ISSUE TITLE
decompose functions change the order of coordinates and features in minibatch

### ISSUE THREAD

**Issue by lilanxiao** [NVIDIA/MinkowskiEngine#225]
Hi, 

I am using the `0.4.3` version. I've noticed that some functions used to decompose the sparse tensor like `SparseTensor.features_at()`, `SparseTensor.coordinates_at()` and `SparseTensor.decomposed_coordinates_and_features`  change the order of coordinates and features within each minibatch.

For example, suppose `st` is a `SparseTensor` and I want to get the first 10 coordinates in the first minibatch. The order of coordinates changes if I call `st.coordinates_at(0)`.
```
st = SparseTensor(coords, feats)
xyz0 = coords[:10, 1:]
xyz1 = st.C[:10, 1:]                      # same result as xyz0
xyz2 = st.coordinates_at(0)[:10, :]          # different as xyz0 and xyz1 
```
In my current work, I want that the decomposed features and coordinates have the same order in each minibatch as the input tensor `coords` and `feats`. Instead of calling the built-in functions, I have to use following method, which seems ugly and slow.

```
feats_minibatch = st.F[ st.C[:, 0] == minibatch_index ]
```
 
Is it an intended behavior to change the order after decomposing? Is there a good way to get the mapping between these two orders?

Thank you very much!

**Comment by chrischoy** [comment]
Have you used the `decomposed_coordinates_and_features`? This would return coordinates and features with the same order.

**Comment by lilanxiao** [comment]
> Have you used the `decomposed_coordinates_and_features`? This would return coordinates and features with the same order.

@chrischoy 
yes, I've tried. This API changes the order too. Please check this code to reproduce:

```
import MinkowskiEngine as me
import torch
import numpy as np
np.random.seed(42)

coords = np.random.randint(size=[100, 3], low=1, high=100)
coords = torch.IntTensor(coords)
feats = np.random.rand(100, 2).astype(np.float32)
feats = torch.from_numpy(feats)
print("input tensor: \n", coords[:3, :])
c, f = me.utils.sparse_collate(coords=[coords, coords], 
                                feats=[feats, feats])
st = me.SparseTensor(coords=c, feats=f)
print("direct access: \n", st.C[:3, 1:])
clist, flist = st.decomposed_coordinates_and_features
print("decomposed: \n", clist[0][:3, :])
```

The result I get is: 
```
input tensor: 
 tensor([[52, 93, 15],
        [72, 61, 21],
        [83, 87, 75]], dtype=torch.int32)
direct access: 
 tensor([[52, 93, 15],
        [72, 61, 21],
        [83, 87, 75]], dtype=torch.int32)
decomposed: 
 tensor([[23, 15, 43],
        [32, 70, 32],
        [42, 77, 51]], dtype=torch.int32)
```

**Comment by zgojcic** [comment]
@lilanxiao: I think what @chrischoy meant was that if one uses  `decomposed_coordinates_and_features` the order of output features and coordinates is consistent (it is still random and different from the input order though).

@chrischoy: maybe you could add a small note about this randomness in the documentation. I encountered a similar problem when I was using `decomposed_features` in combination with the raw coordinates before voxelization and it took me days to find the problem.

**Comment by lilanxiao** [comment]
@zgojcic yeah, I got it. I have done exactly the same thing as you! I mixed the row coordinates and decomposed features too and got totally bad result. I spent days to check my dataloader, model configuration, loss function and some extensions written by myself, wrote tons of test cases and finally realized what caused the problem. :joy:

**Comment by chrischoy** [comment]
The coordinate list does not assume that the coordinates for a specific batch index are contiguous.

[[0, 1]]
[[1, 1]]
[[0, 0]]

is also a plausible coordinate.

I added the function [`decomposition_permutations`](https://github.com/NVIDIA/MinkowskiEngine/blob/5f07586b5094d86c648395f33b9c66b8c85e6766/MinkowskiEngine/MinkowskiTensor.py#L423) to retrieve the permutation used to generate the decomposed coordinates and features.

**Comment by lilanxiao** [comment]
@chrischoy thank you very much for the update! will close the issue.
